python/ray/tests/kuberay/utils.py

```diff
@@ -153,16 +180,54 @@ def get_pod(pod_name_filter: str, namespace: str) -> str:
 
 
 def kubectl_exec(
-    command: List[str], pod: str, namespace: str, container: Optional[str] = None
-) -> None:
+    command: List[str],
+    pod: str,
+    namespace: str,
+    container: Optional[str] = None,
+    return_out: bool = False,
+) -> Optional[str]:
     """kubectl exec the `command` in the given `pod` in the given `namespace`.
     If a `container` is specified, will specify that container for kubectl.
+
+    Args:
+        return_out: If True, stdout will be captured, printed, and returned as a string.
+            Otherwise, stdout will not be redirected and None will be returned.
     """
     container_option = ["-c", container] if container else []
-
-    subprocess.check_call(
+    kubectl_exec_command = (
         ["kubectl", "exec", "-it", pod] + container_option + ["--"] + command
     )
+    if return_out:
+        out = subprocess.check_output(kubectl_exec_command).decode().strip()
+        # Print for debugging convenience.
+        print(out)
+        return out
+    else:
+        subprocess.check_call(kubectl_exec_command)
+        return None
+
+
+def kubectl_exec_python_script(
+    script_name: str,
+    pod: str,
+    namespace: str,
+    container: Optional[str] = None,
+    return_out: bool = False,
+) -> Optional[str]:
+    """
+    Runs a python script in a container via `kubectl exec`.
+    Scripts live in `tests/kuberay/scripts`.
+
+    Args:
+        script_name: The name of a script in tests/kuberay/scripts.
+        return_out: If True, stdout will be redirected to the function's output.
+            Otherwise, stdout will not be redirected and `None` will be returned.
+    """
+    script_path = Path(__file__).resolve().parent / "scripts" / script_name
+    script_string = open(script_path).read()
```

Comment:
This should be done with a with statement or else a file descriptor gets leaked.

Reply:
Will fix that.

The file descriptor is closed upon garbage collection of the Python object.
But this is not part of Python's specification, so it's bad form.

I picked up this bad habit by reading the Ray python code, which does this all over the place...

Reply:
Closed all the file descriptors.